res/controllers/PIONEER-DDJ-ERGO-scripts.js

```diff
@@ -0,0 +1,441 @@
+    if (value === 0x7F) {
+        engine.scratchEnable(deckNumber, 2048, 33+1/3, alpha, beta);
+    } else {
+        engine.scratchDisable(deckNumber);
+    }
+};
+
+// The wheel that actually controls the scratching
+DDJERGO.wheelTurn = function(channel, control, value, status, group) {
+    const deckNumber = script.deckFromGroup(group);
+    //this is because the controller never sends a 0x40 message and hence never stops scratching
+    if (Math.abs(value-0x40)===1) {
+        value = 0x40;
+    }
+    const newValue = value - 64;
+    if (engine.isScratching(deckNumber)) {
+        engine.scratchTick(deckNumber, newValue);
+    } else {
+        engine.setValue(group, "jog", newValue);
+    }
+};
+
+//multiply/divide loop size when autoloop wheel is turned
+DDJERGO.autoloopMultiply = function(channel, control, value, status, group) {
+    const newValue = value-64;
+    if (newValue > 0) {
+        engine.setParameter(group, "loop_halve", 1);
+    }
+    if (newValue < 0) {
+        engine.setParameter(group, "loop_double", 1);
+    }
+};
+
+//this inverts the tempo slider. Didn't seem to work in the xml
+DDJERGO.rate = function(channel, control, value, status, group) {
+    if (control === 0x00) {
+        const newValue = value-64;
+        engine.setValue(group, "rate", -1*newValue/64);
+    }
+};
+
+//change tempo rate in 8% increments (32% max) with shift + keylock button
+DDJERGO.setTempoRange = function(channel, control, value, status, group) {
+    if (value === 0x7F) {
+        let curr = engine.getValue(group, "rateRange");
+        if (curr > 0.31) {
+            curr = 0.08;
+        } else {
+            curr += 0.08;
+        }
+        engine.setValue(group, "rateRange", curr);
+    }
+};
+
+//move quickly through the track when shift + jog wheel touch + turn
+DDJERGO.searchTrack = function(channel, control, value, status, group) {
+    //this is because the controller never sends a 0x40 message and hence never stops searching
+    if (Math.abs(value-0x40)===1) {
+        value = 0x40;
+    }
+    const newValue = value-64;
+    // 3 * val seems to yield the best results
+    engine.setValue(group, "rateSearch", 3*newValue);
+};
+
+//move the current loop by one beat when shift + in/out
+DDJERGO.loopMove = function(channel, control, value, status, group) {
+    if (value === 0x7F) {
+        if (control === 0x4C) {
+            engine.setValue(group, "loop_move", -1);
+        }
+        if (control === 0x4D) {
+            engine.setValue(group, "loop_move", 1);
+        }
+    }
+};
+
+//move the beat markers to match the beats with shift + auto loop
+DDJERGO.adjustGrid = function(channel, control, value, status, group) {
+    let newValue;
+    if (value-64 > 0) { newValue = value-128; } else { newValue = value; }
+    if (newValue > 0) {
+        engine.setValue(group, "beats_translate_later", 1);
+    }
+    if (newValue < 0) {
+        engine.setValue(group, "beats_translate_earlier", 1);
+    }
+};
```

Comment:
I mapped this following the virtual dj mapping as close as possible. I'm not sure if this controller supports long press (hold). Will take a look but that's the why of this combo

Reply:
> I mapped this following the virtual dj mapping as close as possible

Understandable, but there's no reason to mimic the (potentially sub-optimal) behaviour another software.
Let's start from zero and try to make it work good with Mixxx and cover as many use cases as possible.

Reply:
ok so unfortunately ERGO doesn't support long press. Should I proceed with the next implementation then? 

AUTOLOOP turn: resize loop
Shift + AUTOLOOP turn: adjust `beatjump_size`
Shift + In/Out: move loop by `beatjump_size`

This will mean discarding the move beatgrid

Reply:
> ok so unfortunately ERGO doesn't support long press

Hmm, what signals does the AUTOLOOP send on press?
Only press (no release)?
Or press+release immediately?

Reply:
> This will mean discarding the move beatgrid

Well, beatjump is more important for performance than beatgrid move, so I'd rather drop the latter.

Maybe you can squeeze it in elsewhere?
Shift + wheel turn?
Or press Loop_In + LoopOut + touch + turn? With the hand on the platter (not just a finger on the rim) you can move very precisely.